### PR TITLE
Borders

### PR DESCRIPTION
**Pull request by jacobp100**
Currently working on this. Making a PR for initial feedback!

### CODE REVIEW

File: src/node-to-svg.ts [hunk 1]
```diff
@@ -1,30 +1,142 @@
...
+    Top,
+    Right,
+    Bottom,
+    Left
+}
+
+// Done in CSS order
+type Sides<T> = [T, T, T, T]
+
+const applySides = <T>(sides: Sides<T>, value: T | null | undefined, side: Side): Sides<T> => {
```
Comment (orta) [inline src/node-to-svg.ts]: This is a very clever answer to the problem 👍

**Comment by orta** [comment]
This is looking 👍

**Comment by jacobp100** [comment]
@orta Okay, it should be working! Would be great to hear your thoughts! I'll still need to come back to it and do some clean-up here and there.

One thing I noticed in `react-sketchapp` is that they expand their styles when rendering their components ([from here](https://github.com/airbnb/react-sketchapp/blob/master/src/components/View.js) to [flattenStyle](https://github.com/airbnb/react-sketchapp/blob/master/src/stylesheet/index.js), then [expandStyle](https://github.com/airbnb/react-sketchapp/blob/master/src/stylesheet/expandStyle.js)). Would it be worth trying to use this, which would reduce complexity in `component-to-node` and `node-to-svg`?

File: src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
    <g transform='translate(20, 120)'>
-     <rect type="View" fill-opacity="0.1" stroke-width="1" stroke="black" x="20" y="0" width="10" height="10"/>
-     <rect type="View" fill-opacity="0.1" stroke-width="1" stroke="black" x="10" y="20" width="10" height="10"/>
-     <rect type="View" fill-opacity="0.1" stroke-width="1" stroke="black" x="0" y="40" width="10" height="10"/>
+     <rect type="View" x="20" y="0" width="10" height="10" fill="none" stroke="black" stroke-width="0" rx="0" ry="0"/>
+     <rect type="View" x="10" y="20" width="10" height="10" fill="none" stroke="black" stroke-width="0" rx="0" ry="0"/>
+     <rect type="View" x="0" y="40" width="10" height="10" fill="none" stroke="black" stroke-width="0" rx="0" ry="0"/>
    </g>
 
  </g>
```
Comment (orta) [inline src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg]: Something that'll need to be addressed is what to show when there is a lack of  any border settings, almost all of these SVGs are not useful for visual regressions _unless_ you have borders set on each element.

screen shot 2017-10-16 at 4 19 19 pm

File: src/_tests/example_layouts/__snapshots__/_borders.test.tsx-varying-border-radii-dotted.svg [hunk 3]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<svg width="320" height="480" xmlns="http://www.w3.org/2000/svg" version="1.1">
+
+ <path type="View" fill="red" d="M50,0L90,0A10,10 0 0,1 100,10L100,80A20,20 0 0,1 80,100L40,100A40,40 0 0,1 0,60L0,50A50,50 0 0,1 50,0Z"/><path stroke-linecap="round" stroke-dasharray="0, 15" fill="none" stroke="blue" stroke-width="10" d="M50,5L90,5A5,5 0 0,1 95,10L95,80A15,15 0 0,1 80,95L40,95A35,35 0 0,1 5,60L5,50A45,45 0 0,1 50,5Z"/>
+</svg>
```
Comment (orta) [inline src/_tests/example_layouts/__snapshots__/_borders.test.tsx-varying-border-radii-dotted.svg]: Good tests

File: src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg (hunk 2 above)
Comment (jacobp100) [reply src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg]: I’m not sure I follow. In RN, you wouldn’t see anything either.

Comment (jacobp100) [reply src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg]: Is the idea to demonstrate layout?

Comment (orta) [reply src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg]: Yeah, I built this project specifically for visualizing layouts ( see the [VISION file](https://github.com/jest-community/jest-snapshots-svg/blob/master/VISION.md) ) removing that ability basically makes it unusable for what I started it for

Comment (jacobp100) [reply src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg]: Got you! How would you feel about having a wireframe layout mode that shows all boxes, and a visual layout mode that tries to look at close to the actual design as we can?

Comment (orta) [reply src/_tests/__snapshots__/render.test.tsx-handles-some-simple-jsx.svg]: Yep, fine by me 👍

File: src/svg/ViewWireframe.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    "type": "View",
+    "fill": "none",
+    "stroke": "black",
+    "stroke-width": borderWidth,
+  }
+
+  if (sidesEqual(borderRadii)) {
+    const borderRadius = borderRadii[0]
+    // Offset size by half border radius, as RN draws border inside, whereas SVG draws on both sides
+    return $("rect", {
```
Comment (jacobp100) [inline src/svg/ViewWireframe.ts]: @orta It's not hooked up yet, but does this look like it would work?

Comment (orta) [reply src/svg/ViewWireframe.ts]: Yep 👍

File: src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-nested-wireframe.svg [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (jacobp100) [inline src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-nested-wireframe.svg]: @orta How's this? I think before you were keeping the background colour. What do you think is best when background colours are included?

Comment (orta) [reply src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-nested-wireframe.svg]: This is 👍  by me - at some point I'd like to do more work on making it look pretty, but keeping it the same is

Comment (jacobp100) [reply src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-nested-wireframe.svg]: Did you want to keep the background/border colours?

Comment (orta) [reply src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-nested-wireframe.svg]: If it's blank - yeah 👍

File: src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-simple-wireframe-with-background-color.svg [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (jacobp100) [inline src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-simple-wireframe-with-background-color.svg]: @orta Sorry it took so long to update this PR. Is this what you wanted?

Comment (orta) [reply src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-simple-wireframe-with-background-color.svg]: I don't think so, it feels like the  crux of the matter is that we're building two different-ish things:

* You're making a 💯 SVG renderer for a React Component
* I'm aiming something that shows the layout of my React components

For example, in this PR, this SVG went from showing something - to showing nothing:

screen shot 2017-12-06 at 9 32 29 am

Which is basically what will happen to all my tests if I upgrade to this version of jest-snapshots-svg 😺 

I **don't** think these two aims are divergent goals at all, and perhaps the multi-package route is  the right option option: where there is a package for react-to-svg rendering, and the testing lib which can fill in the gaps like "this component has no border or BG, so add a default". 

In the mean time, I'm happy to figure a way to get this in somehow, it's been a long PR and it'd be good to get it in

Comment (jacobp100) [reply src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-simple-wireframe-with-background-color.svg]: I've updated it again. I think this was just a few tests wanting the `{ wireframe: true }` option. Let me know if this works for your use-case! (Sorry again about the delay)

Comment (orta) [reply src/_tests/example_layouts/__snapshots__/_wireframe.test.tsx-simple-wireframe-with-background-color.svg]: No pressure, this looks 👍

**Comment by orta** [comment]
re-ran the build 👍

**Comment by orta** [comment]
sigh, issues building yoga -  I think if we update the dep to 1.7.0 maybe we'll have some luck

**Comment by orta** [comment]
If you rebase, it should go green: https://github.com/jest-community/jest-snapshots-svg/pull/46

**Comment by orta** [comment]
Want to give rebasing a shot?

**Comment by jacobp100** [comment]
Sorry, missed the message. Just updated the PR!

**Comment by orta** [comment]
👍 ace, let's do this

**Pull request merged**